### PR TITLE
feat(scheduler-targets-alpha): `KinesisDataFirehosePutRecord` Target

### LINKED ISSUE #27450: (aws-scheduler-targets-alpha): Add KinesisDataFirehosePutRecord Target
### Describe the feature

Work to support L2 constructs for AWS Scheduler is in progress (https://github.com/aws/aws-cdk/issues/23394). See the approved [RFC](https://github.com/aws/aws-cdk-rfcs/pull/486/files). RFC planned to add 12 templates targets, but only Lambda Invoke is currently implemented (https://github.com/aws/aws-cdk/pull/26575). 

This issue tracks implementation of  `KinesisDataFirehosePutRecord` target to put a record to a Kinesis Firehose.

### Use Case

Customers would like to use templated target `KinesisDataFirehosePutRecord ` to be able to put a record to a Kinesis Firehose on schedule. L2 target construct should grant required permissions to the `AWS Scheduler` to put records to a Kinesis Firehose.  

### Proposed Solution

The proposed solution needs to be adopted to the recent examples of `LambdaInvoke` (https://github.com/aws/aws-cdk/blob/main/packages/%40aws-cdk/aws-scheduler-targets-alpha/lib/lambda-invoke.ts). 

Solution should also include unit and integration tests. 

```ts
export class KinesisDataFirehosePutRecord extends ScheduleTargetBase implements IScheduleTarget {
    constructor(
      private readonly deliveryStream: firehose.CfnDeliveryStream,
      private readonly props: ScheduleTargetBaseProps,
    ) {
      super(props, deliveryStream);
    }
    
    protected addTargetActionToRole(schedule: ISchedule, role: IRole): void {
     // TODO: Check if target and schedule are in the same account and region
     // TODO: Check if target and role are in the same account

     // TODO: Check if L2 grant methods can be used
      role.addToPrincipalPolicy(new iam.PolicyStatement({
        actions: ['firehose:PutRecord'],
        resources: [this.deliverySream.attrArn],
      }));      
    }
  }
```

### Other Information

_No response_

### Acknowledgements

- [X] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### CDK version used

2.99.1

### Environment details (OS name and version, etc.)

MacOs

### PR DESCRIPTION
**Pull request by go-to-k**
This PR adds KinesisDataFirehosePutRecord as a target for an EventBridge scheduler.

Closes #27450.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-scheduler-targets-alpha/README.md [hunk 1]
```diff
@@ -121,3 +122,27 @@ new Schedule(this, 'Schedule', {
   target: new targets.CodeBuildStartBuild(project),
 });
 ```
+
+## Put records to an Amazon Kinesis Data Firehose
+
+Use the `KinesisDataFirehosePutRecord` target to put records to an Amazon Kinesis Data Firehose.
```
Comment (lpizzinidev) [inline packages/@aws-cdk/aws-scheduler-targets-alpha/README.md:128]:
```suggestion
Use the `KinesisDataFirehosePutRecord` target to put a record to a Kinesis Data Firehose delivery stream.
```

File: packages/@aws-cdk/aws-scheduler-targets-alpha/README.md [hunk 2]
```diff
@@ -121,3 +122,27 @@ new Schedule(this, 'Schedule', {
...
+
+The code snippet below creates an event rule with a delivery stream as a target
+called every hour by Event Bridge Scheduler with a custom payload.
+
+```ts
+import * as kinesisfirehose from 'aws-cdk-lib/aws-kinesisfirehose';
+
+declare const deliveryStream: kinesisfirehose.CfnDeliveryStream;
```
Comment (lpizzinidev) [inline packages/@aws-cdk/aws-scheduler-targets-alpha/README.md:136]:
```suggestion
import * as firehose from 'aws-cdk-lib/aws-kinesisfirehose';

declare const deliveryStream: firehose.CfnDeliveryStream;
```

File: packages/@aws-cdk/aws-scheduler-targets-alpha/lib/kinesis-data-firehose-put-record.ts [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+  constructor(
+    private readonly deliveryStream: CfnDeliveryStream,
+    private readonly props: ScheduleTargetBaseProps = {},
+  ) {
+    super(props, deliveryStream.attrArn);
+  }
+
+  protected addTargetActionToRole(schedule: ISchedule, role: IRole): void {
+    if (!sameEnvDimension(this.deliveryStream.stack.region, schedule.env.region)) {
+      throw new Error(`Cannot assign firehose in region ${this.deliveryStream.stack.region} to the schedule ${Names.nodeUniqueId(schedule.node)} in region ${schedule.env.region}. Both the schedule and the firehose must be in the same region.`);
```
Comment (lpizzinidev) [inline packages/@aws-cdk/aws-scheduler-targets-alpha/lib/kinesis-data-firehose-put-record.ts:21]:
```suggestion
      throw new Error(`Cannot assign the Firehose delivery stream in region ${this.deliveryStream.stack.region} to the schedule ${Names.nodeUniqueId(schedule.node)} in region ${schedule.env.region}. Both the schedule and the Firehose delivery stream must be in the same region.`);
```

File: packages/@aws-cdk/aws-scheduler-targets-alpha/lib/kinesis-data-firehose-put-record.ts [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    }
+
+    if (!sameEnvDimension(this.deliveryStream.stack.account, schedule.env.account)) {
+      throw new Error(`Cannot assign firehose in account ${this.deliveryStream.stack.account} to the schedule ${Names.nodeUniqueId(schedule.node)} in account ${schedule.env.region}. Both the schedule and the firehose must be in the same account.`);
```
Comment (lpizzinidev) [inline packages/@aws-cdk/aws-scheduler-targets-alpha/lib/kinesis-data-firehose-put-record.ts:25]:
```suggestion
      throw new Error(`Cannot assign the Firehose delivery stream in account ${this.deliveryStream.stack.account} to the schedule ${Names.nodeUniqueId(schedule.node)} in account ${schedule.env.region}. Both the schedule and the Firehose delivery stream must be in the same account.`);
```

**Review by lpizzinidev — CHANGES_REQUESTED** [review]
Looks good 👍
Just some minor documentation adjustments needed.

**Comment by go-to-k** [comment]
@lpizzinidev 

Thanks! Please check the changes.

**Review by lpizzinidev — APPROVED** [review]
Thanks 👍

**Review by vinayak-kukreja — APPROVED** [review]
Hey @go-to-k, thank you for this contribution.

**Comment by go-to-k** [comment]
@vinayak-kukreja 

Thanks for your approval! I fixed conflicts. Please approve again.

I fixed [the other PR](https://github.com/aws/aws-cdk/pull/27845) the same way, but when one of them gets merged, the other one gets conflicts again. If that happens, I will fix it again.

**Comment by vinayak-kukreja** [comment]
Hey @go-to-k , thank you for fixing the conflicts. The other PR is now merged. Please let me know once you have fixed conflicts for this PR.

**Comment by go-to-k** [comment]
@vinayak-kukreja 

Thanks, I fixed new conflicts in this PR again, please look at this.

**Review by vinayak-kukreja — APPROVED** [review]

**Pull request merged**